File: src/lib/taste.ts
export const AXES = [
  "sweet",
  "salty",
  "sour",
  "umami",
  "spicy",
  "nutty",
] as const;
export const LABELS = ["단맛", "짠맛", "신맛", "감칠맛", "매운맛", "고소함"];
export type Vector = number[];
export type Food = {
  id: string;
  name: string;
  category: string;
  emoji: string;
  description: string;
  vector: Vector;
  color: string;
  vector_version: string;
};
export type Choice = {
  selected: string;
  rejected: string;
  question_id: string;
  sequence: number;
};
export type Context = { address: string; time: string; weather: string };
export type Order = {
  id: string;
  menu_id: string;
  food_id: string;
  vector: Vector;
  context: Context;
  repeat_order: boolean;
  applied: boolean;
  created_at: string;
};
export type Claim = {
  axis: (typeof AXES)[number];
  level: number;
  comparative: boolean;
  evidence: string;
  review_id: string;
};
export type Menu = {
  id: string;
  food_id: string;
  name: string;
  price: number;
  rating: number;
  vector: Vector;
  claims: Claim[];
  source_type: string;
  evidence_count: number;
  description: string;
};
export const VERSION = "taste-v1.0";
export const clip = (n: number) => Math.max(0, Math.min(1, n));
export const distance = (a: Vector, b: Vector) =>
  Math.sqrt(a.reduce((s, v, i) => s + (v - b[i]) ** 2, 0) / 6);
export const match = (a: Vector, b: Vector) => 100 * (1 - distance(a, b));
export function estimate(choices: Choice[], foods: Food[]): Vector {
  const map = new Map(foods.map((f) => [f.id, f.vector]));
  // PRD 8.2: RMS-distance logistic likelihood, beta=8, L2 prior=0.05.
  const loss = (u: Vector) =>
    choices.reduce((total, c) => {
      const selected = map.get(c.selected),
        rejected = map.get(c.rejected);
      if (!selected || !rejected) throw new Error("알 수 없는 음식입니다.");
      const z = 8 * (distance(u, rejected) - distance(u, selected));
      return total + Math.log1p(Math.exp(-z));
    }, 0) +
    0.05 * u.reduce((s, v) => s + (v - 0.5) ** 2, 0);
  let u = Array(6).fill(0.5),
    best = loss(u);
  // Deterministic bounded coordinate descent; avoids an optimization service.
  for (const step of [0.2, 0.1, 0.05, 0.02, 0.01, 0.005, 0.001]) {
    for (let round = 0; round < 70; round++) {
      let improved = false;
      for (let axis = 0; axis < 6; axis++)
        for (const dir of [-1, 1]) {
          const candidate = [...u];
          candidate[axis] = clip(candidate[axis] + dir * step);
          const value = loss(candidate);
          if (value < best - 1e-10) {
            u = candidate;
            best = value;
            improved = true;
          }
        }
      if (!improved) break;
    }
  }
  return u.map((n) => Math.round(n * 10000) / 10000);
}
const COVERAGE = [
  ["bulgogi", "gukbap"],
  ["ramen", "porridge"],
  ["naengmyeon", "udon"],
  ["kimchi", "salad"],
  ["jeyuk", "donkatsu"],
  ["kongguksu", "pho"],
];
export function question(choices: Choice[], foods: Food[]): [Food, Food] {
  const map = new Map(foods.map((f) => [f.id, f]));
  if (choices.length < 6)
    return COVERAGE[choices.length].map((id) => map.get(id)!) as [Food, Food];
  const u = estimate(choices, foods);
  const evidence = Array(6).fill(0);
  for (const c of choices)
    for (let k = 0; k < 6; k++)
      evidence[k] += Math.abs(
        map.get(c.selected)!.vector[k] - map.get(c.rejected)!.vector[k],
      );
  const uncertain = evidence.indexOf(Math.min(...evidence));
  const used = new Set(
    choices.map((c) => [c.selected, c.rejected].sort().join(":")),
  );
  let result: [Food, Food] = [foods[0], foods[1]],
    best = -Infinity;
  for (let a = 0; a < foods.length; a++)
    for (let b = a + 1; b < foods.length; b++) {
      if (used.has([foods[a].id, foods[b].id].sort().join(":"))) continue;
      const contrast = Math.abs(
        foods[a].vector[uncertain] - foods[b].vector[uncertain],
      );
      const ambiguity = Math.abs(
        distance(u, foods[a].vector) - distance(u, foods[b].vector),
      );
      const score = contrast - ambiguity * 0.7;
      if (score > best) {
        best = score;
        result = [foods[a], foods[b]];
      }
    }
  return result;
}
export function currentVector(
  base: Vector,
  orders: Order[],
  context: Context,
): Vector {
  if (!orders.length) return [...base];
  const mean = (rows: Order[]) =>
    AXES.map(
      (_, k) => rows.reduce((sum, o) => sum + o.vector[k], 0) / rows.length,
    );
  const overall = mean(orders),
    output = [...base];
  const specs = [
    ["address", 0.4, 10],
    ["time", 0.3, 12],
    ["weather", 0.2, 20],
  ] as const;
  for (const [key, scale, prior] of specs) {
    if (context[key] === "unknown") continue;
    const rows = orders.filter((o) => o.context[key] === context[key]);
    if (!rows.length) continue;
    const local = mean(rows),
      weight = (scale * rows.length) / (rows.length + prior);
    for (let k = 0; k < 6; k++) output[k] += weight * (local[k] - overall[k]);
  }
  return output.map(clip);
}
export function rank<T extends { id: string; vector: Vector }>(
  items: T[],
  vector: Vector,
) {
  return items
    .map((item) => ({ ...item, score: match(vector, item.vector) }))
    .sort((a, b) => b.score - a.score || a.id.localeCompare(b.id));
}
export function reason(user: Vector, food: Vector) {
  const axes = AXES.map((_, i) => ({ i, d: Math.abs(user[i] - food[i]) }))
    .sort((a, b) => a.d - b.d)
    .slice(0, 2);
  return `${LABELS[axes[0].i]}과 ${LABELS[axes[1].i]}의 강도가 내 취향과 가까워요.`;
}
export function timeSegment(date = new Date()) {
  const h = Number(
    new Intl.DateTimeFormat("en-US", {
      timeZone: "Asia/Seoul",
      hour: "numeric",
      hourCycle: "h23",
    }).format(date),
  );
  return h >= 6 && h < 11
    ? "breakfast"
    : h < 14 && h >= 11
      ? "lunch"
      : h >= 14 && h < 17
        ? "afternoon"
        : h >= 17 && h < 21
          ? "dinner"
          : "late-night";
}
export function restaurantVectors(
  base: Vector,
  menus: { id: string; claims: Claim[] }[],
) {
  return menus.map((menu) => ({
    id: menu.id,
    vector: AXES.map((axis, k) => {
      const claims = menu.claims.filter((c) => c.axis === axis);
      if (!claims.length) return base[k];
      const raw = (cs: Claim[]) =>
        cs.reduce((s, c) => s + c.level * (c.comparative ? 1.25 : 1), 0) /
        cs.length;
      const peers = menus
        .map((m) => m.claims.filter((c) => c.axis === axis))
        .filter((cs) => cs.length)
        .map(raw);
      if (peers.length < 2) return base[k];
      const value = raw(claims),
        less = peers.filter((v) => v < value).length,
        equal = peers.filter((v) => v === value).length;
      const percentile = (less + (equal - 1) / 2) / (peers.length - 1);
      const n = new Set(claims.map((c) => c.review_id)).size;
      return clip(base[k] + (2 * (percentile - 0.5) * 0.15 * n) / (n + 10));
    }),
  }));
}

File: api/core/taste.py
import math
from datetime import datetime, timezone, timedelta
from typing import List, Dict, Any, Tuple, Optional

TASTE_AXES = ["sweet", "salty", "sour", "umami", "spicy", "nutty"]
AXIS_LABELS = ["단맛", "짠맛", "신맛", "감칠맛", "매운맛", "고소함"]

DEFAULT_COVERAGE_PAIRS = [
    ["bulgogi", "gukbap"],
    ["ramen", "porridge"],
    ["naengmyeon", "udon"],
    ["kimchi", "salad"],
    ["jeyuk", "donkatsu"],
    ["kongguksu", "pho"],
]

def clip(x: float, low: float = 0.0, high: float = 1.0) -> float:
    return max(low, min(high, x))

def distance(a: List[float], b: List[float]) -> float:
    """RMS distance across 6 dimensions."""
    if len(a) != 6 or len(b) != 6:
        raise ValueError("Vectors must have exactly 6 dimensions.")
    variance = sum((a[i] - b[i]) ** 2 for i in range(6)) / 6.0
    return math.sqrt(variance)

def match_score(a: List[float], b: List[float]) -> float:
    """Match percentage: 100 * (1 - RMS distance)."""
    return 100.0 * (1.0 - distance(a, b))

def estimate_u_base(choices: List[Dict[str, Any]], food_vectors: Dict[str, List[float]]) -> List[float]:
    """
    Estimate U_base using Bounded Coordinate Descent.
    Loss = -sum(log(P(actual))) + 0.05 * ||U - 0.5||^2
    where z = 8 * (d(U, rejected) - d(U, selected))
    """
    if not choices:
        return [0.5] * 6

    def compute_loss(u: List[float]) -> float:
        total = 0.0
        for c in choices:
            selected_id = c["selected"]
            rejected_id = c["rejected"]
            if selected_id not in food_vectors or rejected_id not in food_vectors:
                raise ValueError(f"Unknown food ID: {selected_id} or {rejected_id}")
            sel = food_vectors[selected_id]
            rej = food_vectors[rejected_id]
            z = 8.0 * (distance(u, rej) - distance(u, sel))
            if z >= 40.0:
                total += math.exp(-z)
            elif z <= -40.0:
                total += -z
            else:
                total += math.log1p(math.exp(-z))
        reg = 0.05 * sum((u[k] - 0.5) ** 2 for k in range(6))
        return total + reg

    u = [0.5] * 6
    best_loss = compute_loss(u)

    steps = [0.2, 0.1, 0.05, 0.02, 0.01, 0.005, 0.001]
    for step in steps:
        for _ in range(70):
            improved = False
            for axis in range(6):
                for direction in (-1.0, 1.0):
                    candidate = list(u)
                    candidate[axis] = clip(candidate[axis] + direction * step)
                    cand_loss = compute_loss(candidate)
                    if cand_loss < best_loss - 1e-10:
                        u = candidate
                        best_loss = cand_loss
                        improved = True
            if not improved:
                break

    return [round(val, 4) for val in u]

def get_next_question(
    choices: List[Dict[str, Any]],
    foods: List[Dict[str, Any]],
    coverage_pairs: Optional[List[List[str]]] = None
) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    """
    Returns the next pair of foods for Pairwise Onboarding.
    Q1-Q6: Coverage pairs.
    Q7-Q10: Adaptive uncertainty reduction.
    """
    coverage = coverage_pairs or DEFAULT_COVERAGE_PAIRS
    food_map = {f["id"]: f for f in foods}

    if len(choices) < len(coverage):
        pair_ids = coverage[len(choices)]
        return food_map[pair_ids[0]], food_map[pair_ids[1]]

    vector_map = {f["id"]: f["vector"] for f in foods}
    u_temp = estimate_u_base(choices, vector_map)

    # Calculate evidence per axis
    evidence = [0.0] * 6
    for c in choices:
        sel_vec = vector_map[c["selected"]]
        rej_vec = vector_map[c["rejected"]]
        for k in range(6):
            evidence[k] += abs(sel_vec[k] - rej_vec[k])

    uncertain_axis = evidence.index(min(evidence))

    used_pairs = {":".join(sorted([c["selected"], c["rejected"]])) for c in choices}

    best_pair: Optional[Tuple[Dict[str, Any], Dict[str, Any]]] = None
    best_utility = -float("inf")

    for i in range(len(foods)):
        for j in range(i + 1, len(foods)):
            pair_key = ":".join(sorted([foods[i]["id"], foods[j]["id"]]))
            if pair_key in used_pairs:
                continue

            fa = foods[i]["vector"]
            fb = foods[j]["vector"]

            target_contrast = abs(fa[uncertain_axis] - fb[uncertain_axis])
            ambiguity = abs(distance(u_temp, fa) - distance(u_temp, fb))
            utility = target_contrast - 0.7 * ambiguity

            if utility > best_utility:
                best_utility = utility
                best_pair = (foods[i], foods[j])

    if best_pair is None:
        return foods[0], foods[1]

    return best_pair

def current_vector(
    base_vector: List[float],
    orders: List[Dict[str, Any]],
    context: Dict[str, str]
) -> List[float]:
    """
    Calculates U_current = U_base + Place + Time + Weather with shrinkage.
    """
    if not orders:
        return list(base_vector)

    def mean_vector(order_rows: List[Dict[str, Any]]) -> List[float]:
        n = len(order_rows)
        return [sum(o["vector"][k] for o in order_rows) / n for k in range(6)]

    overall_mean = mean_vector(orders)
    output = list(base_vector)

    specs = [
        ("address", 0.4, 10.0),
        ("time", 0.3, 12.0),
        ("weather", 0.2, 20.0),
    ]

    for key, scale, prior in specs:
        ctx_val = context.get(key, "unknown")
        if ctx_val == "unknown":
            continue

        matched_orders = [o for o in orders if o.get("context", {}).get(key) == ctx_val]
        if not matched_orders:
            continue

        local_mean = mean_vector(matched_orders)
        n = len(matched_orders)
        weight = (scale * n) / (n + prior)

        for k in range(6):
            output[k] += weight * (local_mean[k] - overall_mean[k])

    return [clip(round(v, 4)) for v in output]

def time_segment(dt: Optional[datetime] = None) -> str:
    """Classifies time of day in Asia/Seoul (UTC+9)."""
    if dt is None:
        seoul_tz = timezone(timedelta(hours=9))
        dt = datetime.now(seoul_tz)
    elif dt.tzinfo is None:
        seoul_tz = timezone(timedelta(hours=9))
        dt = dt.replace(tzinfo=seoul_tz)

    hour = dt.hour
    if 6 <= hour < 11:
        return "breakfast"
    elif 11 <= hour < 14:
        return "lunch"
    elif 14 <= hour < 17:
        return "afternoon"
    elif 17 <= hour < 21:
        return "dinner"
    else:
        return "late-night"

def generate_reason(user_vec: List[float], food_vec: List[float]) -> str:
    """Generates user-facing match explanation from closest taste axes."""
    diffs = sorted(
        [(abs(user_vec[i] - food_vec[i]), i) for i in range(6)],
        key=lambda x: x[0]
    )
    first_axis = AXIS_LABELS[diffs[0][1]]
    second_axis = AXIS_LABELS[diffs[1][1]]
    return f"{first_axis}과 {second_axis}의 강도가 내 취향과 가까워요."
